pygfx/utils/compute.py

```diff
@@ -0,0 +1,254 @@
+                bindings.append(
+                    {
+                        "binding": index,
+                        "resource": wgpu_object.create_view(
+                            usage=wgpu.TextureUsage.STORAGE_BINDING
+                        ),
+                    }
+                )
+            else:
+                raise RuntimeError(f"Unexpected resource: {resource}")
+        return bindings
+
+    def dispatch(self, nx, ny=1, nz=1):
+        """Dispatch the workgroups, i.e. run the shader."""
+        nx, ny, nz = int(nx), int(ny), int(nz)
+
+        # Reset
+        self._changed = False
+
+        # Get device
+        if self._device is None:
+            self._shader_module = None
+            self._device = gfx.renderers.wgpu.Shared.get_instance().device
+        device = self._device
+
+        # Compile the shader
+        if self._shader_module is None:
+            self._pipeline = None
+            self._shader_module = device.create_shader_module(
+                label=self._label, code=self._wgsl
+            )
+
+        # Get the pipeline object
+        if self._pipeline is None:
+            self._bind_group = None
+            self._pipeline = device.create_compute_pipeline(
+                label=self._label,
+                layout="auto",
+                compute={
+                    "module": self._shader_module,
+                    "entry_point": self._entry_point,
+                    "constants": self._constants,
+                },
+            )
+
+        # Get the bind group object
+        if self._bind_group is None:
+            bind_group_layout = self._pipeline.get_bind_group_layout(0)
+            bindings = self._get_bindings_from_resources()
+            self._bind_group = device.create_bind_group(
+                label=self._label, layout=bind_group_layout, entries=bindings
+            )
+
+        # Make sure that all used resources have a wgpu-representation, and are synced
+        for resource, _ in self._resources.values():
+            if isinstance(resource, gfx.Resource):
+                gfx.renderers.wgpu.engine.update.update_resource(resource)
+
+        t0 = time.perf_counter()
+
+        # Start!
+        command_encoder = device.create_command_encoder(label=self._label)
+
+        # Maybe clear some buffers
+        for resource, clear in self._resources.values():
+            if clear:
+                command_encoder.clear_buffer(self._get_native_resource(resource))
+
+        # Do the compute pass
+        compute_pass = command_encoder.begin_compute_pass()
+        compute_pass.set_pipeline(self._pipeline)
+        compute_pass.set_bind_group(0, self._bind_group)
+        compute_pass.dispatch_workgroups(nx, ny, nz)
+        compute_pass.end()
+
+        # Submit!
+        device.queue.submit([command_encoder.finish()])
+
+        # Timeit
+        if self._report_time:
+            device._poll_wait()  # wait for the GPU to finish
```

Comment:
Should poll wait be outside the if statement?

Reply:
No, the poll is only needed so that the timing measures the actual time the GPU is working. Without the poll, the GPU is doing its thing while the CPU can also proceed.